File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char_index = {}  # To store the index of each character in the current substring
        max_length = 0
        start = 0  # Start of the current substring

        for end in range(len(s)):
            # If the character is already in the current substring, update the start position
            if s[end] in char_index and char_index[s[end]] >= start:
                start = char_index[s[end]] + 1

            # Update the index of the current character
            char_index[s[end]] = end

            # Update the maximum length if needed
            max_length = max(max_length, end - start + 1)

        return max_length

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {

        HashSet<Character> str = new HashSet<>();

        int slow = 0 ;
        int fast = 0 ;
        int longest = 0 ;

        while(fast < s.length()){
            if(str.add(s.charAt(fast))){
                longest = Math.max(longest , fast - slow + 1 );
                fast++ ;
            }else {
                str.remove(s.charAt(slow));
                slow++ ;
            }
        }

        return longest ;

    }
}
